### ISSUE TITLE
Synthetic temporal neurons via sparse autoencoders

### ISSUE THREAD

**Issue by justinshenk** [justinshenk/temporal-awareness#4]
## Motivation

Current approach uses linear probes to detect temporal encoding, but probes are:
- **Distributed**: The probe direction is a combination of many features
- **Opaque**: We can't say "this neuron fires for long-term thinking"
- **Potentially confounded**: Linear combinations might detect shortcuts

Sparse autoencoders (SAEs) can decompose activations into **monosemantic features** - individual directions that correspond to single concepts.

**Key question**: Are there individual features that activate specifically for temporal horizons?

## Background: Sparse Autoencoders

SAEs learn to reconstruct activations using a sparse overcomplete basis:
```
h = encoder(x)  # Sparse high-dim representation
x' = decoder(h) # Reconstruction
```

With L1 sparsity penalty, each feature tends to be monosemantic (one concept per feature).

Recent work:
- [Anthropic - Scaling Monosemanticity](https://transformer-circuits.pub/2024/scaling-monosemanticity/)
- [Towards Monosemanticity](https://transformer-circuits.pub/2023/monosemantic-features/)

### Phase 1: Train SAE on Temporal Activations

1. Extract activations from Layer 10 (best semantic layer per probe validation)
2. Train SAE on mixed corpus including temporal prompts
3. Identify features that correlate with temporal labels

```python
# Pseudocode
activations = extract_layer_10_activations(temporal_prompts + general_prompts)
sae = SparseAutoencoder(input_dim=768, hidden_dim=768*8, sparsity=0.01)
sae.fit(activations)

# Find temporal features
for feature_idx in range(sae.hidden_dim):
    feature_activations = sae.encode(temporal_activations)[:, feature_idx]
    correlation = spearman(feature_activations, temporal_labels)
    if abs(correlation) > 0.3:
        print(f"Feature {feature_idx}: r={correlation}")
```

### Phase 2: Analyze Temporal Features

For each candidate temporal feature:
1. Find max-activating prompts
2. Visualize feature across prompt distribution
3. Test if feature is monosemantic (one concept) or polysemantic (multiple)

**Monosemanticity test**: Do max-activating prompts share only temporal semantics?

### Phase 3: Causal Intervention

1. Zero out temporal feature during generation
2. Amplify temporal feature
3. Measure effect on:
   - Model's stated time horizon
   - Downstream task behavior
   - Probe predictions

```python
def ablate_feature(model, sae, feature_idx, prompt):
    """Generate with specific feature ablated."""
    def hook(module, input, output):
        h = sae.encode(output)
        h[:, feature_idx] = 0  # Ablate
        return sae.decode(h)
    
    model.layer_10.register_forward_hook(hook)
    return model.generate(prompt)
```

### Phase 4: Compare to Linear Probes

| Metric | Linear Probe | SAE Features |
|--------|--------------|--------------|
| Interpretability | Low (distributed) | High (monosemantic) |
| Accuracy | 84% (Layer 10) | TBD |
| Causal validity | Unknown | Testable via ablation |
| Feature count | 1 direction | Possibly multiple features |

**Hypothesis**: Multiple SAE features encode different *aspects* of temporal reasoning:
- Feature A: Urgency/deadline pressure
- Feature B: Long-term consequences
- Feature C: Planning vs. reacting
- Feature D: Reversibility

## Tools & Resources

- **SAE implementations**: 
  - [mats-sae-training](https://github.com/jbloomAus/mats_sae_training)
  - [transformer-lens](https://github.com/neelnanda-io/TransformerLens) with SAE support
- **Compute**: SAE training is more expensive than probes (~hours on GPU)
- **Pre-trained SAEs**: Check if GPT-2 SAEs already exist (Anthropic, EleutherAI)

## Success Criteria

1. **Find ≥1 monosemantic temporal feature** with:
   - Correlation > 0.3 with temporal labels
   - Max-activating prompts sharing temporal theme
   - Causal effect when ablated

2. **Improve interpretability**: Can we describe what the feature detects in plain language?

3. **Discovery**: Do we find temporal sub-categories (urgency, consequences, planning) that probes miss?

## Open Questions

- Do temporal features emerge naturally, or only in temporal training data?
- Is there a "time horizon spectrum" feature, or discrete short/long features?
- How do SAE features relate to the probe direction?
- Can we find features for temporal *uncertainty* (distinct from horizon)?

## References

- Cunningham et al. 2023 - Sparse Autoencoders Find Highly Interpretable Features in Language Models
- Bricken et al. 2023 - Towards Monosemanticity (Anthropic)
- Templeton et al. 2024 - Scaling Monosemanticity (Anthropic)

## Connection to Current Work

- **Probe validation** ✅ confirms Layer 10 encodes temporal semantics
- **SPD decomposition** decomposes parameters, SAEs decompose activations
- Both aim for interpretable components, different granularity
